Correct Unix Socket logic to correctly identify production run times from development.

A user is having an issue connecting to MySQL on the local App Engine dev server: https://stackoverflow.com/q/50705839/1701388

The root cause is diagnosed here: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/3136

In short, Google Cloud Tools for Java set `GAE_RUNTIME` to match the production env when running the local dev server: https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/444. As the code of `SocketFactory` below suggests, it always uses the Unix domain socket if `GAE_RUNTIME` is set. This is especially problematic on Windows.

```java
    // This env will be set by GAE OR set manually if using Cloud SQL Proxy
    String runtime = System.getenv("GAE_RUNTIME");

    if (runtime == null || runtime.isEmpty()) {  // Use standard SSL (direct connection)
      this.socket = SslSocketFactory.getInstance().create(instanceName);
    } else { // Use Unix Socket
      logger.info("Using GAE Unix Sockets");
      UnixSocketAddress socketAddress = new UnixSocketAddress(
          new File(CloudSqlPrefix + instanceName));
      this.socket = UnixSocketChannel.open(socketAddress).socket();
    }
```

Here is what I think is reasonable:
- If `GAE_ENV=localdev`, don't use the Unix domain socket.
or
- Use the Unix socket only when it can find the corresponding socket file under `/cloudsql/`.

Maybe the latter is better, because at least on Linux, users have an option to create a Unix socket locally with the Cloud SQL proxy.

BTW, is there any immediate workaround for the user to make a connection from the local GAE dev server?

About `SslSocketFactory`, I just replaced `com.google.cloud.sql.mysql.SslSocketFactory` with `com.google.cloud.sql.core.SslSocketFactory` in the following connection URL, but it gives me an error, probably because `SslSocketFactory` doesn't have a 0-arg constructor.

```java
    String serverURL = "jdbc:mysql://google/mysql?"
        + "cloudSqlInstance=your-gcp-project-id:us-east1:your-mysql-instance-id"
        + "&socketFactory=com.google.cloud.sql.core.SslSocketFactory"
        + "&user=root&password=does-not-matter"
        + "&useSSL=false";

      conn = DriverManager.getConnection(serverURL);
```

```
Caused by: java.lang.InstantiationException: com.google.cloud.sql.core.SslSocketFactory
	at java.lang.Class.newInstance(Class.java:427)
	at com.mysql.jdbc.MysqlIO.createSocketFactory(MysqlIO.java:3328)
	... 46 more
Caused by: java.lang.NoSuchMethodException: com.google.cloud.sql.core.SslSocketFactory.<init>()
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.newInstance(Class.java:412)
	... 47 more
```

--- a/connector-j-5/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java
+++ b/connector-j-5/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java
@@ -47,18 +47,25 @@ public class SocketFactory implements com.mysql.jdbc.SocketFactory {
         "cloudSqlInstance property not set. Please specify this property in the JDBC URL or "
             + "the connection Properties with value in form \"project:region:instance\"");
 
-    logger.info(String.format("Connecting to Cloud SQL instance [%s].", instanceName));
+    // gaeEnv="standard" indicates standard instances
+    // runEnv="Production" indicates production instances
+    String gaeEnv = System.getenv("GAE_ENV");
+    String runEnv = System.getProperty("com.google.appengine.runtime.environment");
+    // Custom env variable for forcing unix socket
+    Boolean forceUnixSocket = System.getenv("CLOUD_SQL_FORCE_UNIX_SOCKET") != null;
 
-    // This env will be set by GAE OR set manually if using Cloud SQL Proxy
-    String runtime = System.getenv("GAE_RUNTIME");
-
-    if (runtime == null || runtime.isEmpty()) {  // Use standard SSL (direct connection)
-      this.socket = SslSocketFactory.getInstance().create(instanceName);
-    } else { // Use Unix Socket
-      logger.info("Using GAE Unix Sockets");
+    // If running on GAE Standard, connect with unix socket
+    if (forceUnixSocket || "standard".equals(gaeEnv) && "Production".equals(runEnv)) {
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via unix socket.", instanceName));
       UnixSocketAddress socketAddress = new UnixSocketAddress(
           new File(CloudSqlPrefix + instanceName));
       this.socket = UnixSocketChannel.open(socketAddress).socket();
+    } else {
+      // Default to SSL Socket
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via ssl socket.", instanceName));
+      this.socket = SslSocketFactory.getInstance().create(instanceName);
     }
     return this.socket;
   }

--- a/connector-j-6/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java
+++ b/connector-j-6/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java
@@ -47,18 +47,25 @@ public class SocketFactory implements com.mysql.cj.api.io.SocketFactory {
             "cloudSqlInstance property not set. Please specify this property in the JDBC URL or "
                     + "the connection Properties with value in form \"project:region:instance\"");
 
-    logger.info(String.format("Connecting to Cloud SQL instance [%s].", instanceName));
+    // gaeEnv="standard" indicates standard instances
+    // runEnv="Production" indicates production instances
+    String gaeEnv = System.getenv("GAE_ENV");
+    String runEnv = System.getProperty("com.google.appengine.runtime.environment");
+    // Custom env variable for forcing unix socket
+    Boolean forceUnixSocket = System.getenv("CLOUD_SQL_FORCE_UNIX_SOCKET") != null;
 
-    // This env will be set by GAE OR set manually if using Cloud SQL Proxy
-    String runtime = System.getenv("GAE_RUNTIME");
-
-    if (runtime == null || runtime.isEmpty()) {  // Use standard SSL (direct connection)
-      this.socket = SslSocketFactory.getInstance().create(instanceName);
-    } else { // Use Unix Socket
-      logger.info("Using GAE Unix Sockets");
+    // If running on GAE Standard, connect with unix socket
+    if (forceUnixSocket || "standard".equals(gaeEnv) && "Production".equals(runEnv)) {
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via unix socket.", instanceName));
       UnixSocketAddress socketAddress = new UnixSocketAddress(
           new File(CloudSqlPrefix + instanceName));
       this.socket = UnixSocketChannel.open(socketAddress).socket();
+    } else {
+      // Default to SSL Socket
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via ssl socket.", instanceName));
+      this.socket = SslSocketFactory.getInstance().create(instanceName);
     }
     return this.socket;
   }

--- a/connector-j-8/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java
+++ b/connector-j-8/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java
@@ -22,7 +22,6 @@ import com.google.cloud.sql.core.SslSocketFactory;
 import com.mysql.cj.protocol.ServerSession;
 import com.mysql.cj.protocol.SocketConnection;
 
-import java.io.Closeable;
 import java.io.File;
 import java.io.IOException;
 import java.net.Socket;
@@ -51,21 +50,29 @@ public class SocketFactory implements com.mysql.cj.protocol.SocketFactory {
             "cloudSqlInstance property not set. Please specify this property in the JDBC URL or "
                     + "the connection Properties with value in form \"project:region:instance\"");
 
-    logger.info(String.format("Connecting to Cloud SQL instance [%s].", instanceName));
+    // gaeEnv="standard" indicates standard instances
+    // runEnv="Production" indicates production instances
+    String gaeEnv = System.getenv("GAE_ENV");
+    String runEnv = System.getProperty("com.google.appengine.runtime.environment");
+    // Custom env variable for forcing unix socket
+    Boolean forceUnixSocket = System.getenv("CLOUD_SQL_FORCE_UNIX_SOCKET") != null;
 
-    // This env will be set by GAE OR set manually if using Cloud SQL Proxy
-    String runtime = System.getenv("GAE_RUNTIME");
-
-    if (runtime == null || runtime.isEmpty()) {  // Use standard SSL (direct connection)
-      this.socket = SslSocketFactory.getInstance().create(instanceName);
-    } else { // Use Unix Socket
-      logger.info("Using GAE Unix Sockets");
+    // If running on GAE Standard, connect with unix socket
+    if (forceUnixSocket || "standard".equals(gaeEnv) && "Production".equals(runEnv)) {
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via unix socket.", instanceName));
       UnixSocketAddress socketAddress = new UnixSocketAddress(
           new File(CloudSqlPrefix + instanceName));
       this.socket = UnixSocketChannel.open(socketAddress).socket();
+    } else {
+      // Default to SSL Socket
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via ssl socket.", instanceName));
+      this.socket = SslSocketFactory.getInstance().create(instanceName);
     }
     return this.socket;
   }
+
 
   // Cloud SQL sockets always use TLS and the socket returned by connect above is already TLS-ready. It is fine
   // to implement these as no-ops.

--- a/postgres/src/main/java/com/google/cloud/sql/postgres/SocketFactory.java
+++ b/postgres/src/main/java/com/google/cloud/sql/postgres/SocketFactory.java
@@ -52,19 +52,25 @@ public class SocketFactory extends javax.net.SocketFactory {
 
   @Override
   public Socket createSocket() throws IOException {
-    logger.info(String.format("Connecting to Cloud SQL instance [%s].", instanceName));
+    // gaeEnv="standard" indicates standard instances
+    // runEnv="Production" indicates production instances
+    String gaeEnv = System.getenv("GAE_ENV");
+    String runEnv = System.getProperty("com.google.appengine.runtime.environment");
+    // Custom env variable for forcing unix socket
+    Boolean forceUnixSocket = System.getenv("CLOUD_SQL_FORCE_UNIX_SOCKET") != null;
 
-    // This env will be set by GAE OR set manually if using Cloud SQL Proxy
-    String runtime = System.getenv("GAE_RUNTIME");
-
-    if (runtime == null || runtime.isEmpty()) {  // Use standard SSL (direct connection)
-      return SslSocketFactory.getInstance().create(instanceName);
+    // If running on GAE Standard, connect with unix socket
+    if (forceUnixSocket || "standard".equals(gaeEnv) && "Production".equals(runEnv)) {
+      logger.info(String.format(
+          "Connecting to Cloud SQL instance [%s] via unix socket.", instanceName));
+      UnixSocketAddress socketAddress = new UnixSocketAddress(
+          new File(CloudSqlPrefix + instanceName + PostgreSqlSufix));
+      return UnixSocketChannel.open(socketAddress).socket();
     }
-    logger.info("GAE Unix Sockets");
-    UnixSocketAddress socketAddress = new UnixSocketAddress(
-        new File(CloudSqlPrefix + instanceName + PostgreSqlSufix));
-    UnixSocket socket = UnixSocketChannel.open(socketAddress).socket();
-    return socket;
+    // Default to SSL Socket
+    logger.info(String.format(
+        "Connecting to Cloud SQL instance [%s] via ssl socket.", instanceName));
+    return SslSocketFactory.getInstance().create(instanceName);
   }
 
   @Override

